tensorboard/compat/tensorflow_stub/io/gfile.py

```diff
@@ -0,0 +1,620 @@
+    """Provides filesystem access to S3."""
+
+    def __init__(self):
+        if not boto3:
+            raise ImportError("boto3 must be installed for S3 support.")
+
+    def bucket_and_path(self, url):
+        """Split an S3-prefixed URL into bucket and path."""
+        url = compat.as_str_any(url)
+        if url.startswith("s3://"):
+            url = url[len("s3://"):]
+        idx = url.index("/")
+        bucket = url[:idx]
+        path = url[(idx + 1):]
+        return bucket, path
+
+    def exists(self, filename):
+        """Determines whether a path exists or not."""
+        client = boto3.client("s3")
+        bucket, path = self.bucket_and_path(filename)
+        r = client.list_objects(Bucket=bucket, Prefix=path, Delimiter="/")
+        if r.get("Contents") or r.get("CommonPrefixes"):
+            return True
+        return False
+
+    def remove(self, filename):
+        """Deletes the file located at 'filename'."""
+        s3 = boto3.resource("s3")
+        bucket, path = self.bucket_and_path(filename)
+        s3.Object(bucket, path).delete()
+
+    def rmtree(self, dirname):
+        """Deletes everything under dirname recursively."""
+        s3 = boto3.resource("s3")
+        bucket, path = self.bucket_and_path(dirname)
+        s3_bucket = s3.Bucket(bucket)
+        for obj in s3_bucket.objects.filter(Prefix=path):
+            obj.delete()
+
+    def read(self, filename, binary_mode=False, size=None, offset=None):
+        """Reads contents of a file to a string.
+
+        Args:
+            filename: string, a path
+            binary_mode: bool, read as binary if True, otherwise text
+            size: int, number of bytes or characters to read, otherwise
+                read all the contents of the file from the offset
+            offset: int, offset into file to read from, otherwise read
+                from the very beginning
+
+        Returns:
+            Subset of the contents of the file as a string or bytes.
+        """
+        s3 = boto3.resource("s3")
+        bucket, path = self.bucket_and_path(filename)
+        args = {}
+        if size is not None or offset is not None:
+            if offset is None:
+                offset = 0
+            args['Range'] = 'bytes={}-{}'.format(offset, offset + size)
+        stream = s3.Object(bucket, path).get(**args)['Body'].read()
+        if binary_mode:
+            return bytes(stream)
+        else:
+            return stream.decode('utf-8')
+
+    def glob(self, filename):
+        """Returns a list of files that match the given pattern(s)."""
+        # Only support prefix with * at the end and no ? in the string
+        star_i = filename.find('*')
+        quest_i = filename.find('?')
+        if quest_i >= 0:
+            raise NotImplementedError(
+                "{} not supported by compat glob".format(filename))
+        if star_i != len(filename) - 1:
+            # Just return empty so we can use glob from directory watcher
```

Comment:
I can see your point here. However, we want to use this for our internal distributed filesystem as well and doing a check for S3 in GetLogdirSubdirectories wouldn't be sufficient. Can we put a TODO here instead of making the larger change?